### ISSUE TITLE
Review the new DVC sharing experiments workflow and unify it with VS Code

### ISSUE THREAD

**Issue by shcheklein** [iterative/vscode-dvc#3574]
With https://github.com/iterative/dvc/issues/9074 and https://github.com/iterative/studio/issues/5050 DVC introduced a way to connect to Studio and a workflow that sends experiments to Studio.

We need to review and unify/make it compatible:

- [x] Token management https://github.com/iterative/dvc/issues/9265 (#3768)
- [x] Auto-sharing mode - does it contradict or not with existing DVC workflow? - no, leave as is
- [x] Drop custom client implementation (#3701)
- [x] Consolidate share actions into one - https://github.com/iterative/vscode-dvc/issues/3574#issuecomment-1505741392 (#3771 + #3781)
- [x] Improvement: Sharing multiple experiments action https://github.com/iterative/vscode-dvc/issues/3574#issuecomment-1513236062 (#3792)
- [x] When a user is not already connected to Studio prompt with a toast to connect on exp push https://github.com/iterative/vscode-dvc/issues/3574#issuecomment-1513949115 (#3781)
- [x] Add instructions to live share outside of the extension context (https://github.com/iterative/vscode-dvc/issues/3574#issuecomment-1521609421) - not needed (https://github.com/iterative/vscode-dvc/issues/3574#issuecomment-1529304891)
- [x] Sharing experiments. Can we detect that it's already being shared? Show a link?

**Comment by dberenbaum** [comment]
> Can we detect that it's already being shared?  Show a link?

We could fetch from the git remote or use `dvc exp list` to see what's been pushed. Do we need anything specific to Studio?

I'm a bit behind on the Studio efforts to have more shareable links, but I think we can follow it in https://github.com/iterative/studio/issues/5419 and https://github.com/iterative/studio/issues/5150.

> * Auto-sharing mode - does it contradict or not with existing DVC workflow?

I guess it means this:

<img width="372" alt="Screenshot 2023-03-29 at 3 42 58 PM" src="https://user-images.githubusercontent.com/2308172/228650823-0fdbe88a-c67d-4e04-be5a-78af29354e29.png">

I don't think it contradicts. Does it do anything besides add the `STUDIO_TOKEN` to experiment actions @mattseddon? I think it's needed since there's no other obvious way to enable this behavior when working in the extension.

> * Do we still need to use our own client implementation or not?

I hope not.

**Comment by dberenbaum** [comment]
> * Sharing experiments. Can we detect that it's already being shared? Show a link?

I don't know if we have a good way to detect what experiments have been live shared. For exp refs, we could use `dvc exp list` to check the git remote. If they exist there, there should be a record in Studio, although it's possible it hasn't been retrieved/parsed yet by Studio.

I would add an item to consolidate and unify the sharing options. Here's what I see today:

<img width="207" alt="Screenshot 2023-04-12 at 2 16 34 PM" src="https://user-images.githubusercontent.com/2308172/231548449-49a33e9f-da63-49c6-a420-04edace364f7.png">

Overall, I think we can have a single "Share" action that relies on `dvc exp push` now and drop the existing different ways to share. This will hopefully make sharing more straightforward.

Now Studio should show any experiments that are pushed to the Git remote via `dvc exp push`, so we shouldn't need to rely on the live metrics API anymore. `dvc exp push` will notify Studio if `STUDIO_TOKEN` is set or if the token is available in `studio.token` in the DVC config (depends on https://github.com/iterative/dvc/pull/9257; we will likely recommend using `dvc config --global studio.token xxx` for CLI users).

`dvc exp push` will push the cache by default. Not sure if we need an option to disable that, but we should at least be aware that it could make pushing slow.

**Comment by shcheklein** [comment]
> Sharing experiments. Can we detect that it's already being shared? Show a link?

I think DVCLive creates some special signal file with a process pid? Can we utilize it cc @mattseddon @daavoo . 

On the "share" actions. I agree, @dberenbaum  we can add item to the list to consolidate this and simplify. We still keep an apply action though, right? So that I can make locally a commit out of an experiment at least. We just don't send these commits, neither we try to do `dvc push`, etc.

**Comment by dberenbaum** [comment]
> We still keep an apply action though, right? So that I can make locally a commit out of an experiment at least. We just don't send these commits, neither we try to do `dvc push`, etc.

Agreed, we can keep local actions like "Apply to workspace" and "Create new branch," I just don't think we need to combine them with the sharing action.

**Comment by mattseddon** [comment]
> I would add an item to consolidate and unify the sharing options. Here's what I see today:
> 
> <img alt="Screenshot 2023-04-12 at 2 16 34 PM" width="207" src="https://user-images.githubusercontent.com/2308172/231548449-49a33e9f-da63-49c6-a420-04edace364f7.png">
> 
> Overall, I think we can have a single "Share" action that relies on `dvc exp push` now and drop the existing different ways to share. This will hopefully make sharing more straightforward.
> 
> Now Studio should show any experiments that are pushed to the Git remote via `dvc exp push`, so we shouldn't need to rely on the live metrics API anymore. `dvc exp push` will notify Studio if `STUDIO_TOKEN` is set or if the token is available in `studio.token` in the DVC config (depends on [iterative/dvc#9257](https://github.com/iterative/dvc/pull/9257); we will likely recommend using `dvc config --global studio.token xxx` for CLI users).

I'm happy to drop `Commit and Share` + `Branch and Share` and update `Share to Studio` to use `exp push` and label it with simply `Share`. My one question would be "how do we promote Studio as a sharing option"? Currently, with the custom implementation of the Studio client and the command's label, we can do a few things:

1. Promote Studio as a sharing option.
2. We can check to see if there is a Studio token set for the extension and redirect to the setup page if there isn't one present.
3. If we receive a 401 we can notify the user that they need to update their token.

Should we drop this logic altogether and not worry about the edge cases? 
How do we show users that sharing to Studio is an option? 

It feels like we need an API that we can use at startup to check whether or not the user has a valid token and if not prompt them to add one/update. Otherwise, we can't provide much help to users once they leave the happy path.

> dvc exp push will push the cache by default. Not sure if we need an option to disable that, but we should at least be aware that it could make pushing slow.

From testing out with the demo project moving from the old API to using `exp push` blew the time out to share an experiment from 1s to ~90s, after the initial push this dropped to 14s. It would be good to be able to reduce that time in any way possible.

**Comment by dberenbaum** [comment]
> I'm happy to drop `Commit and Share` + `Branch and Share` and update `Share to Studio` to use `exp push` and label it with simply `Share`. My one question would be "how do we promote Studio as a sharing option"? Currently, with the custom implementation of the Studio client and the command's label, we can do a few things:

Sorry, I'm getting confused by all the different ways to share. Do you mean how do we promote live sharing to Studio, or any kind of sharing to Studio?


> It feels like we need an API that we can use at startup to check whether or not the user has a valid token and if not prompt them to add one/update.

Not sure I follow all the suggestions, but can we start simpler by checking whether the user has a token at all and assume it's valid if it exists? I think checking the token validity is less of a priority for now.

> From testing out with the demo project moving from the old API to using `exp push` blew the time out to share an experiment from 1s to ~90s, after the initial push this dropped to 14s.

Are you testing with checkpoints?

**Comment by mattseddon** [comment]
> Sorry, I'm getting confused by all the different ways to share. Do you mean how do we promote live sharing to Studio, or any kind of sharing to Studio?

Any kind of sharing to Studio. How do we highlight this as a collaboration option? 

Note: I can see from https://github.com/iterative/dvc.org/pull/4470 that if a user has a token present in their config then they can live share experiments but not sure what the default behaviour is. 

> Not sure I follow all the suggestions, but can we start simpler by checking whether the user has a token at all and assume it's valid if it exists? I think checking the token validity is less of a priority for now.

I think we need a Studio API that we can check auth details against but this should be fixed/made obsolete by https://github.com/iterative/studio/issues/5158. 

A couple of follow-up questions:

- Can I use `dvc config studio.token` to check if a user has a Studio token in their config? (if yes I can use it as an environment variable if "live sharing" is enabled) 
- When setting the token into a user's config should I set it as `--global` or `--system`? 

If users have already set this up I think we should still prompt them to add again as this makes it explicit that we are saving it in their DVC config (rather than expending effort migrating the secret).

> Are you testing with checkpoints?

In this instance no, checkpoints were dropped on the branch that I was testing with.

**Comment by dberenbaum** [comment]
Thanks @mattseddon! A couple of not directly related thoughts before responding to your questions/comments.

Looking through the Studio onboarding from VS Code, do we need the "sign in" button? The "get token" button will take them to the same place if not signed in.

To make sharing more useful, can we also include an action to share after selecting multiple experiments (it looks like you can only share one at a time now)?

> Any kind of sharing to Studio. How do we highlight this as a collaboration option?

When choosing to share, we could prompt them with something like `Connect to Studio to see and manage shared experiments` if they aren't connected already (with an option to stop showing the prompt). Open to other ideas here.

> * Can I use `dvc config studio.token` to check if a user has a Studio token in their config? (if yes I can use it as an environment variable if "live sharing" is enabled)

Yes and yes.

> * When setting the token into a user's config should I set it as `--local`, `--global` or `--system`? (my guess is `--global`)

Yes `--global` sounds good to me.

> If users have already set this up I think we should still prompt them to add again as this makes it explicit that we are saving it in their DVC config (rather than expending effort migrating the secret).

Sounds good.



> From testing out with the demo project moving from the old API to using `exp push` blew the time out to share an experiment from 1s to ~90s, after the initial push this dropped to 14s. It would be good to be able to reduce that time in any way possible.

We probably need an option to share with/without the cache, whether that is two separate sharing actions or whether there's some other way (prompt, settings option, etc.).

**Comment by mattseddon** [comment]
> Looking through the Studio onboarding from VS Code, do we need the "sign in" button? The "get token" button will take them to the same place if not signed in.

We can do this. I was just trying to give a linear/defined process for the user to follow. We can drop the sign-in button. The user may just have to use some brain power.

> To make sharing more useful, can we also include an action to share after selecting multiple experiments (it looks like you can only share one at a time now)?

This is correct. I will enable it through the multi-select context menu. When using `exp push` with multiple experiments is push parallelised? If not I will push individually and update the progress notification each time an experiment is pushed.

> When choosing to share, we could prompt them with something like Connect to Studio to see and manage shared experiments if they aren't connected already (with an option to stop showing the prompt). Open to other ideas here.

Sounds reasonable. I'll add buttons to a toast that let the user: 
1. open the setup page with the studio section focused
2. sets a config option that means they won't be asked again

> We probably need an option to share with/without the cache, whether that is two separate sharing actions or whether there's some other way (prompt, settings option, etc.).

After considering the implications a bit closer I think we should stick with the default, to begin with.

I'll update the description. Please LMK if I have missed anything 🙏🏻.

**Comment by dberenbaum** [comment]
> Drop the sign-in button from Studio setup page

IMO it's simpler with fewer buttons, but Studio won't redirect users to the token if they aren't already logged in, so it could also cause confusion. Up to you what to do.

> When using `exp push` with multiple experiments is push parallelised? If not I will push individually and update the progress notification each time an experiment is pushed.

Yes, at least I think there's some optimization when pushing the cache for multiple experiments, so I would suggest doing it that way. The final output should have the info you want (what was pushed to both git and the cache), but it won't be streamed with each individual experiment, and there's currently no machine-readable output format.



> > We probably need an option to share with/without the cache, whether that is two separate sharing actions or whether there's some other way (prompt, settings option, etc.).
> 
> After considering the implications a bit closer I think we should stick with the default, to begin with.

Okay. Would like to hear more of your thoughts on this since I agree it will be slow and maybe not what people want when they are only interested in things like metrics and plots.

**Comment by mattseddon** [comment]
> > Drop the sign-in button from Studio setup page
> 
> IMO it's simpler with fewer buttons, but Studio won't redirect users to the token if they aren't already logged in, so it could also cause confusion. Up to you what to do.

Should be fixed by https://github.com/iterative/studio/issues/5715. Maybe we can wait for that.

> > After considering the implications a bit closer I think we should stick with the default, to begin with.
> 
> Okay. Would like to hear more of your thoughts on this since I agree it will be slow and maybe not what people want when they are only interested in things like metrics and plots.

I was thinking about this from a pessimistic/"what could go wrong" with someone's workflow point of view. If they push the cache by default then at least everything will be re-creatable/shared and there shouldn't be any surprises later on. IIUC this helps to solve the "I forgot the push" issue.

**Comment by dberenbaum** [comment]
> If they push the cache by default then at least everything will be re-creatable/shared and there shouldn't be any surprises later on. IIUC this helps to solve the "I forgot the push" issue.

> When choosing to share, we could prompt them with something like `Connect to Studio to see and manage shared experiments` if they aren't connected already (with an option to stop showing the prompt). Open to other ideas here.

If we are going to have a prompt when sharing, we could include a checkbox for whether to share the cache, which could be checked by default.

**Comment by mattseddon** [comment]
> If we are going to have a prompt when sharing, we could include a checkbox for whether to share the cache, which could be checked by default.

This would be on the settings page, right? We could start with it in the "Studio" section but I think we would want to move it to the new "Remote" section once we get to it. Maybe we can delay until that is added? Up to you.

**Comment by dberenbaum** [comment]
> This would be on the settings page, right?

I was thinking it could be part of this:

> Sounds reasonable. I'll add buttons to a toast that let the user:
> 
>     1. open the setup page with the studio section focused
> 
>     2. sets a config option that means they won't be asked again

---

> Maybe we can delay until that is added?

Yup, we can start without it.

**Comment by dberenbaum** [comment]
Another small item to add: When running a notebook or other no-pipeline workflow, the checbox for sharing new experiments live won't do anything. We could show instructions like "For experiments run outside of this extension, set the DVC_STUDIO_TOKEN environment variable to share experiments live."

**Comment by mattseddon** [comment]
[Q] Do we want an "unshare" option for experiments? I.e implement `dvc exp remove -g origin <name>`

**Comment by dberenbaum** [comment]
> [Q] Do we want an "unshare" option for experiments? I.e implement `dvc exp remove -g origin <name>`

Good question. IMHO it can wait for a follow up later since it's likely not that commonly needed, it's more natural to do from Studio, and it may be complex/hard to describe.

**Comment by dberenbaum** [comment]
A couple takeaways from the AH discussion:

1. @dmpetrov Made a point about distinguishing between all the different options we have in the workflow today. For the `dvc exp push` action, what do you think about explicitly using "push" instead of "share" for the action name? I think it is more explicit and specific, giving more clarity that it's related to git push/dvc push and separate from "live" sharing.
2. He also pointed out that in the video I shared, dropping into the terminal doesn't look good, and I think it's a blocker for putting out a video on YT. To get something working well enough for a YT video, we at minimum need an action in the extension that does `dvc exp push`, so WDYT about prioritizing this first? Ideally it will use the token, but that's not a blocker because reloading the page in Studio will be enough to show the pushed experiments.

**Comment by mattseddon** [comment]
@dberenbaum I can rename to push. Most of the work is done. I will make a release early next week (probably your Monday). PR is here: https://github.com/iterative/vscode-dvc/pull/3781 

The chain underneath will cover 
- Token management
- Drop custom client implementation
- Consolidate share actions into one
- When a user is not already connected to Studio prompt with a toast to connect on exp push

**Comment by mattseddon** [comment]
> Another small item to add: When running a notebook or other no-pipeline workflow, the checbox for sharing new experiments live won't do anything. We could show instructions like "For experiments run outside of this extension, set the DVC_STUDIO_TOKEN environment variable to share experiments live."

I've just realised that this isn't possible. Any attempt made by a user to run an experiment through the extension will be redirected to the "Add a stage" command:

https://user-images.githubusercontent.com/37993418/235397816-3baef59a-fab2-458c-9dda-3a1d44adcff4.mov

I don't think I need to take any further action on this one.

**Comment by dberenbaum** [comment]
> > Another small item to add: When running a notebook or other no-pipeline workflow, the checbox for sharing new experiments live won't do anything. We could show instructions like "For experiments run outside of this extension, set the DVC_STUDIO_TOKEN environment variable to share experiments live."
> 
> I've just realised that this isn't possible. Any attempt made by a user to run an experiment through the extension will be redirected to the "Add a stage" command:

Sorry for the lack of clarity. I meant here:

<img width="529" alt="Screenshot 2023-05-01 at 10 07 47 AM" src="https://user-images.githubusercontent.com/2308172/235464148-1da0e4b3-0bf0-4481-ab76-e2c7eb193dc1.png">

However, maybe it's just confusing to mention it. I think it's fine to leave it out for now.

**Comment by mattseddon** [comment]
> > > Another small item to add: When running a notebook or other no-pipeline workflow, the checbox for sharing new experiments live won't do anything. We could show instructions like "For experiments run outside of this extension, set the DVC_STUDIO_TOKEN environment variable to share experiments live."
> > 
> > 
> > I've just realised that this isn't possible. Any attempt made by a user to run an experiment through the extension will be redirected to the "Add a stage" command:
> 
> Sorry for the lack of clarity. I meant here:
> 
> <img alt="Screenshot 2023-05-01 at 10 07 47 AM" width="529" src="https://user-images.githubusercontent.com/2308172/235464148-1da0e4b3-0bf0-4481-ab76-e2c7eb193dc1.png">
> 
> However, maybe it's just confusing to mention it. I think it's fine to leave it out for now.

In order to get to this screen the user would have to go through the process of generating/adding a Studio token, do you think they would do this without having a pipeline setup? Even then they would have to run their training script through the terminal and not the extension. If I was in that position I wouldn't expect that checking that box would have any effect on running my training script but not sure whether or not that would apply to users.

**Comment by mattseddon** [comment]
The remaining two action items:

> - [ ] Sharing experiments. Can we detect that it's already being shared? Show a link?
> - [ ] Drop the sign-in button from Studio setup page 


are blocked by https://github.com/iterative/studio/issues/5150 & https://github.com/iterative/studio/issues/5715 respectively.

There may be some follow-up on https://github.com/iterative/vscode-dvc/pull/3781#issuecomment-1529864199 / https://github.com/iterative/vscode-dvc/pull/3781#issuecomment-1530804732 but if needed that should be straightforward enough.

**Comment by dberenbaum** [comment]
> * Sharing experiments. Can we detect that it's already being shared? Show a link?

@shcheklein WDYT about making this a separate story/issue? Related to your comment [here](https://github.com/iterative/vscode-dvc/pull/3781#issuecomment-1532007207), it feels like there is more to this. How do we show what has been shared between live Studio metrics, Git refs, and DVC remotes? How do we keep it in sync?

**Comment by mattseddon** [comment]
Remaining item is related to https://github.com/iterative/studio/pull/5417 & https://github.com/iterative/studio/issues/5150

**Comment by mattseddon** [comment]
> Sharing experiments. Can we detect that it's already being shared? Show a link?

Generating a link is now possible after https://github.com/iterative/studio/pull/6149. I don't think that we can detect that an experiment has already been shared without introducing the significant overhead of checking the remote. What I do think we can do is run `dvc exp push origin <exp-name> --no-cache` when a user requests a shareable link. 

One idea that I had is to add a "link" icon into the cell row actions like this:

<img width="898" alt="image" src="https://github.com/iterative/vscode-dvc/assets/37993418/b9fe1af5-d52a-440e-9a91-9805807a74c5">

(positioning up for debate)

On-click the user will have a link copied to their clipboard if they are connected to Studio (this could take a little bit of time because we need to push the experiment). If they are not connected then we can prompt them to connect.

WDYT? Is this good enough to get us unblocked?

**Comment by mattseddon** [comment]
We could also call `exp list origin` with a list of revisions at the same time that we call `exp show` (with the same watcher trigger). This would avoid having to "generate links" but the call seems fairly expensive. I guess we would only enable the link icon for rows that are shareable, the rest would show the icon as disabled.

**Comment by mattseddon** [comment]
Raised https://github.com/iterative/studio/issues/6249 in Studio.

**Comment by shcheklein** [comment]
Reviewing the whole thing one more. And I feel lost tbh.

- @dberenbaum now `save_dvc_exp=True` is ignored in VS Code since we use `dvc exp run`.  Even if "share new experiments" enabled it doesn't send the final result. What was the reason to ignore it? What is the expected way of sending them now in VS Code?
- what does the checkbox mean now?

**Comment by shcheklein** [comment]
Unblocking this. Let's go back and review one more time this.

**Comment by mattseddon** [comment]
@shcheklein behaviour seems related to https://github.com/iterative/dvc/issues/9349. I have seen this mentioned elsewhere but struggling to find the issue.

**Comment by dberenbaum** [comment]
Not sure I follow your comment overall @shcheklein. Are you lost about the whole experience? Is your confusion only about pushing the final result to the git and dvc remotes, or is there something else?

**Comment by shcheklein** [comment]
> Are you lost about the whole experience?

Yep, I feel like that. I'm not sure I understand why `save_dvc_exp` is not working anymore. I'm also not sure I understand what does checkbox do now. Also, yes, what is the expected workflow in VS Code to share the experiments at the end.

**Comment by shcheklein** [comment]
> In what way is it not working? What were you expecting it to do, and what is happening instead?

I think it was pushing the experiment at the end, no? (so that I can see and create a PR in Studio). Now I see an incomplete experiment in Studio. 

> If so, it means you will automatically get live updates in studio for any experiments you run.

those that don't have dvclive enabled at all? do I need to use `dvc exp run` for that? or even from a script that updates the table in the workspace? how do we name it then?

> Currently, the workflow is to use the Push action to do this.

okay, I see. So, what is the recommended way to use it - do we want to use and recommend `save_dvc_exp` in VS Code?

**Comment by dberenbaum** [comment]
Maybe you are thinking of `DVC_EXP_AUTO_PUSH`? This was previously used for checkpoints but we discussed bringing it back for this in https://github.com/iterative/dvc/issues/9295.

`save_dvc_exp=True` does not push anything. It will only make sure each dvclive run saves an experiment locally at the tend (if vs code is running the experiment, it doesn't do anything since it uses `dvc exp run` to save the experiment). See https://dvc.org/doc/dvclive/live#parameters.

My take - a good and easy step would be to recommend the push action at the end of the experiment in both CLI and and VS Code. Doing it automatically has some hurdles: it will be slow (making the experiment hang at the end) and it will break if there's anything wrong with the git or dvc remote connections (incorrect setup, expired credentials, etc.). Both of these may make the experiment workflow feel broken, so I'd rather suggest users do it themselves.

**Comment by shcheklein** [comment]
Okay, let me step back, and think. What really bothers me (no solutions yet, sorry,  just to reflect on why it all doesn't fit into my head when I'm trying that):

- I have to know about push overall. It's not common in the other systems. It is also amplified by the combination of: seeing an experiment in Studio already, on top of that having the checkbox and flag, and yes lack of any extra notifications.
- combinatorial complexity - dvclive is set or not, save exp is set or not, checkbox is set or not, etc - too many possibilities. I was not able to predict the behavior. I'm not able to start with a default that fits.
- warning of `exp_save` being ignored mislead me - should I drop it from the code since it's ignored? It's ignored so I lack some functionality / need to replace it with something?

----------

Some practical thoughts:

- it seems `save_exp` should become a default. And only if ppl want to disable it they go for it. That would simplify the happy path, docs, etc. And warning can be triggered only when False is set explicitly and it's being ignored (you are getting a different behavior)
- `push` - I have no idea how to improve this besides some messages. It's unfortunately inherently confusing thing in the whole architecture - two (or three) types of experiments, etc. We should be clear at least about their state in UI then. Don't know how to explain this better. Any ideas? 
- checkbox - we have to put a message around it if this is about VS Code itself pushing any experiments w/o any help from DVClive

Any other ideas on how to simplify this further?

**Comment by dberenbaum** [comment]
Discussed with @shcheklein. I'm not going to summarize everything here, but the most important action item we agreed on was to think of better ways to make the push action feel smoother, easier, more automatic, etc. In particular, we discussed if pushing in the background can make it feel less intrusive even if it's generally a slow operation. For example, can we check what experiments have not yet been pushed and push them automatically at some regular interval?

**Comment by shcheklein** [comment]
Related to this:

https://discord.com/channels/485586884165107732/1129143778805035008/1129344919195111454

> I noticed that dvc exp run warned me that the save_dvc_exp=True setting would be ignored.

> When I run dvc exp run (via the VS Code extension), there is no data/train/dvclive/dvc.yaml being generated, though, at least not after the validation step and model checkpoints. This means that the VS Code extension doesn't show me any plots to monitor training (for which so far I still use TensorBoard as it's more reliable).

**Comment by dberenbaum** [comment]
> When I run dvc exp run (via the VS Code extension), there is no data/train/dvclive/dvc.yaml being generated, though, at least not after the validation step and model checkpoints. This means that the VS Code extension doesn't show me any plots to monitor training (for which so far I still use TensorBoard as it's more reliable).

That sounds like a bug

**Comment by shcheklein** [comment]
@shcheklein to create a separate story / ticket for the async push and exp status in the experiments table.

**Comment by shcheklein** [comment]
Overall, action items are:

- Make `dvc_exp_save` default. It's pending a tactical decision on when it should be done on the `dvclive`
- Removing warning that "`dvc_exp_save` is ignored". Depends tactically on the previous item cc @dberenbaum 
- Improve UX /copy around checkbox to make it more explicit
- Show experiments status in the table (pushed vs not). Can in the upcoming column "Studio link".
- Push experiments and their data async to help people sharing the experiments

**Comment by dberenbaum** [comment]
> * Show experiments status in the table (pushed vs not). Can in the upcoming column "Studio link".

Discussed with @pmrowla who suggested that it might be faster to call `git ls-remote origin 'refs/exps/*'` than to use `dvc exp list`.

**Comment by mattseddon** [comment]
Proposed design:

https://github.com/iterative/vscode-dvc/assets/37993418/c1c35cee-48ac-48a3-89c3-20261f6f4fe3

Pushed status can fill the gap taken by the running spinner. When the experiment has not been pushed we will show an upload icon in the style of the plot icon checkbox. If the experiment is uploaded we will show a cloud icon with a tooltip that let's the user know the experiment has been pushed.

**Comment by dberenbaum** [comment]
Thanks @mattseddon! I like it. A few minor questions:

Can we go back to a spinner or something similar while the push is happening?

Can we coordinate on the iconography with https://github.com/iterative/studio/issues/6870? I think the cloud is a good idea since `dvc exp push` by default pushes to both the git and dvc remotes, but would be good to make sure we have some consistency.

> Note: We are going to struggle to keep what we display in the table synced with the remote data. E.g when (anyone) runs `dvc exp remove -g origin <exp-name>` we have no event/info to trigger a table update.

Will it sync eventually, like when someone pushes another experiment using the extension?

**Comment by shcheklein** [comment]
Thanks @mattseddon for the quick iteration! Should we already introduce a separate column for these states? Where will we put a link to Studio for example?

**Comment by mattseddon** [comment]
> Can we go back to a spinner or something similar while the push is happening?

Yes, we can do that.

> Can we coordinate on the iconography with https://github.com/iterative/studio/issues/6870? I think the cloud is a good idea since dvc exp push by default pushes to both the git and dvc remotes, but would be good to make sure we have some consistency.

Yes, I will keep an eye on that. For native icons choices are limited to [these ones](https://microsoft.github.io/vscode-codicons/dist/codicon.html). Maybe in Studio, we could show a cloud with a slash through it to signify that it is not uploaded (pushed) yet.

> Will it sync eventually, like when someone pushes another experiment using the extension?

Currently, it will sync as soon as there is a change in the workspace. This falls down when `dvc exp remove -g origin <exp-name>` is run from the terminal (no update is shown in the table).

> Thanks @mattseddon for the quick iteration! Should we already introduce a separate column for these states? Where will we put a link to Studio for example?

I considered this and decided that because we are always fighting for space this would be the best option. As there is nothing to show for commits we would be wasting space there as well. 
The Studio link can replace the cloud icon for completed experiments (under the correct circumstances). For running experiments get Studio link can appear in the context menu. WDYT?

**Comment by shcheklein** [comment]
> The Studio link can replace the cloud icon for completed experiments (under the correct circumstances). For running experiments get Studio link can appear in the context menu. WDYT?

Hmm, I don't see how that could happen to be honest.

I would as usual err on the side of being super obvious and look decent vs optimizing things to look good. E.g. if it means three columns with distinct obvious names - I would even go for that. At least initially.

> For running experiments get Studio link can appear in the context menu. WDYT?

My take - that would be confusing - when fields change they semantics.

> As there is nothing to show for commits we would be wasting space there as well.

They also could show Studio link, GH link, their status (we could check in background data eventually).

**Comment by mattseddon** [comment]
> Hmm, I don't see how that could happen to be honest.

Link icon:

<img width="898" alt="image" src="https://github.com/iterative/vscode-dvc/assets/37993418/b9fe1af5-d52a-440e-9a91-9805807a74c5">

**Comment by shcheklein** [comment]
My take - that's too crowded, also we miss making it explicit (e.g. Studio column name would make the purpose clear).

**Comment by mattseddon** [comment]
@dberenbaum @shcheklein is this the correct view/state that we want a shareable link to point to:

<img width="1727" alt="image" src="https://github.com/iterative/vscode-dvc/assets/37993418/51cafcd6-7ad8-4332-ac90-00bdc3e39940">

**Comment by shcheklein** [comment]
I don't think we need it to be selected (checkbox). I think it's better to open plots in this case tbh (since we don't have any other way to highlight the row when it open on the FE, or do we @iterative/studio-frontend ?)

Checkbox enables these actions at the top? (and hides the Plots button, etc). May be it's even fine to keep checkbox selected, but can we then show more actions? 

some ideas, Matt, I think it's totally fine as a first iteration! thanks 🙏

**Comment by mvshmakov** [comment]
> I think it's better to open plots in this case tbh

👍🏻 

We've recently split the `selected` elements from `active` ones. Selected elements correspond to the checkboxes, and activated ones open the plots. Both can be persisted in the URL.

> Checkbox enables these actions at the top?

Yes, they do.

> May be it's even fine to keep checkbox selected, but can we then show more actions?

Not really, there is not enough real estate in the row. We'll have to squash them aggressively. 

If that is acceptable, I'd prefer sticking to the `activated` rows.

**Comment by mattseddon** [comment]
A couple of questions from me:

- Should live experiments have plots that are visible in the Studio UI?
- Do all views/projects have plots? If not do we know the percentage that doesn't have plots?

**Comment by shcheklein** [comment]
> Should live experiments have plots that are visible in the Studio UI?

Yes, in most cases.

> Do all views/projects have plots? If not do we know the percentage that don't have plots?

No, some projects don't and it's a good point. Means that it's better to show the table. Also show the Plots button (so, no checkbox?). 

> If that is acceptable, I'd prefer sticking to the activated rows.

what are the `activated` rows?

**Comment by mvshmakov** [comment]
> what are the activated rows?

<details><summary>Activated row (plot button)</summary>
<p>

<img width="1700" alt="Screenshot 2023-08-18 at 13 58 15" src="https://github.com/iterative/vscode-dvc/assets/28868855/64f94a28-b09f-4e05-99d0-8abc212e8f4b">

</p>
</details> 

<details><summary>Selected row (checkbox)</summary>
<p>

<img width="1701" alt="Screenshot 2023-08-18 at 13 58 37" src="https://github.com/iterative/vscode-dvc/assets/28868855/357d3200-22bd-4eb6-b246-b358859665a9">

</p>
</details>

<details><summary>Activated AND selected row (plot button AND checkbox)</summary>
<p>

<img width="1698" alt="Screenshot 2023-08-18 at 13 58 49" src="https://github.com/iterative/vscode-dvc/assets/28868855/5c84a74b-7ee7-4fa1-a68f-5c8f920a575c">

</p>
</details>

**Comment by shcheklein** [comment]
Okay, I guess let's do checkbox and if there are plots we can do a leap of faith and show them I guess. It's a bit suboptimal tbh. Checkbox are not meant for this use case (they are about making actions on an experiment). We have a way to highlight an experiment (for a second at the moment). I wonder if can / should utilize that instead?

**Comment by mattseddon** [comment]
> Okay, I guess let's do checkbox and if there are plots we can do a leap of faith and show them I guess. It's a bit suboptimal tbh. Checkbox are not meant for this use case (they are about making actions on an experiment). We have a way to highlight an experiment (for a second at the moment). I wonder if can / should utilize that instead?

Coming back to the original suggestion question quickly:

![image](https://github.com/iterative/vscode-dvc/assets/37993418/5341a2af-6638-4a97-b79c-1c986bdb20df)

The reason for using the checkbox here was to apply the show only selected commits filter:

![image](https://github.com/iterative/vscode-dvc/assets/37993418/92c4ead3-cdae-40f0-b728-8f3705a73776)

This means we can share a URL of the form:

`https://studio.iterative.ai/user/{user_id}/projects/{view_id}?showOnlySelected=1&experimentReferences={exp_ref}&activeExperimentReferences={exp_ref}%3Aprimary`

Whoever receives the link will be able to:

1. Compare the experiment with the baseline.
2. Easily view plots if they exist.

I suggested this view because even though it is not optimal I felt that it provides value whilst also eliminating edge cases (e.g. no plots).

I should have explained this, to begin with.

**Comment by mattseddon** [comment]
Getting closer to finishing this now with https://github.com/iterative/studio/pull/7332 ready for review and a rough prototype done on the extension side. 

I have encountered an issue with the proposed links getting bounced if no plots are available for a record. For this reason, I'm going to resort to the following links:

`${this.baseUrl}?showOnlySelected=1&experimentReferences=${sha}` for completed & pushed experiments
`${this.baseUrl}?showOnlySelected=1&liveExperiments=${baselineSha}%3A${name}` for live experiments

`Copy Studio Link` will be available for as a context menu option for all experiments that exist on the user's instance of Studio.

LMK if you have questions regarding any of the above.

🙏🏻

**Comment by jellebouwman** [comment]
> LMK if you have questions regarding any of the above.

> I have encountered an issue with the proposed links getting bounced if no plots are available for a record. For this reason, I'm going to resort to the following links:

Thanks for the writeup @mattseddon! Does this mean a link like https://studio.iterative.ai/user/jellebouwman/projects/demo-bank-customer-churn-esgpd8e876?showOnlySelected=1&panels=plots%2C&commits=67521d6d2acc16218e4cf6378b73732b14754140&activeCommits=67521d6d2acc16218e4cf6378b73732b14754140%3Aprimary (note `activeCommits`) breaks the URL for you?

IMO - we should try to let you trigger the plots for a rev and should fail gracefully (keep the rest of the ProjectURL working) if the rev doesn't have plots. Let me know if that's the case right now, and I can try to work on a fix for you.

**Comment by mattseddon** [comment]
> Thanks for the writeup @mattseddon! Does this mean a link like https://studio.iterative.ai/user/jellebouwman/projects/demo-bank-customer-churn-esgpd8e876?showOnlySelected=1&panels=plots%2C&commits=67521d6d2acc16218e4cf6378b73732b14754140&activeCommits=67521d6d2acc16218e4cf6378b73732b14754140%3Aprimary (note activeCommits) breaks the URL for you?

If there are no plots for a live experiment then I've seen its link get redirected to the main view page. Here are a couple of examples:

- https://studio.iterative.ai/user/mattseddon/projects/vscode-dvc-demo-ynm6t3jxdx?showOnlySelected=1&liveExperiments=37894efde22e96f184ba2531d4a436b74552b731%3Arummy-genu&activeLiveExperiments=37894efde22e96f184ba2531d4a436b74552b731%3Arummy-genu%3Aprimary (works)
-  https://studio.iterative.ai/user/mattseddon/projects/vscode-dvc-demo-ynm6t3jxdx?showOnlySelected=1&liveExperiments=37894efde22e96f184ba2531d4a436b74552b731%3Apagan-eric&activeLiveExperiments=37894efde22e96f184ba2531d4a436b74552b731%3Apagan-eric%3Aprimary (redirected to https://studio.iterative.ai/user/mattseddon/projects/vscode-dvc-demo-ynm6t3jxdx)

Seems safer to provide a link that doesn't get redirected.

**Comment by mattseddon** [comment]
Final demo for this one. 

Shareable links are provided for all experiments found on Studio (live or pushed) via the experiments table context menu:

https://github.com/iterative/vscode-dvc/assets/37993418/25f0bc17-06e9-4b00-9af5-90a999e5403d

**Comment by shcheklein** [comment]
Thanks, Matt. Looks great.
